src/many-types/src/blockchain.rs

```diff
@@ -157,3 +157,52 @@ impl<'d, C> Decode<'d, C> for SingleTransactionQuery {
         result
     }
 }
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum RangeBlockQuery {
+    Height(CborRange<u64>),
+    Time(CborRange<Timestamp>),
+}
+
+// ; A block query over a range of height or time. This cannot be a hash or
+// ; specific height (use `blockchain.block` for specific height/hash).
+// range-block-query =
+//     ; Height range.
+//     { 1 => range<uint> }
+//     ; Time value or time range.
+//     / { 2 => range<time> }
+impl<C> Encode<C> for RangeBlockQuery {
+    fn encode<W: Write>(&self, e: &mut Encoder<W>, _: &mut C) -> Result<(), Error<W::Error>> {
+        match &self {
+            RangeBlockQuery::Height(range) => e.map(1)?.u8(1)?.encode(range)?,
+            RangeBlockQuery::Time(range) => e.map(1)?.u8(2)?.encode(range)?,
+        };
+        Ok(())
+    }
+}
+
+impl<'d, C> Decode<'d, C> for RangeBlockQuery {
+    fn decode(d: &mut Decoder<'d>, _: &mut C) -> Result<Self, decode::Error> {
+        let mut indefinite = false;
+        let key = match d.map()? {
```

Comment:
You could use a `d.map_iter().next()` to skip some code. And it would make it easier to check that there's only one key.

Reply:
I'll need your Rust-fu on this one.